File: src/game/save/startingGear.ts
import type { EquippedStack, InventoryStack, PlayerSave } from './types'
import {
  RETIRED_FISHING_NET_ITEM_ID,
  STARTING_HUNTING_TOOL_ID,
  WEAPON_TOOL_SLOT_ID,
} from './types'

function hasItem(save: PlayerSave, itemId: string): boolean {
  if (save.inventory.some((stack) => stack.itemId === itemId && stack.quantity > 0)) {
    return true
  }
  return Object.values(save.equipment.slots).some(
    (stack) => stack?.itemId === itemId && stack.quantity > 0,
  )
}

/** Ensure the demo hunting Net is owned and equipped when the tool slot is free. */
export function ensureStartingHuntingTool(save: PlayerSave): PlayerSave {
  const slots: Record<string, EquippedStack | null> = { ...save.equipment.slots }
  let inventory = save.inventory.map((stack) => ({ ...stack }))
  const weapon = slots[WEAPON_TOOL_SLOT_ID]

  if (!hasItem({ ...save, inventory, equipment: { slots } }, STARTING_HUNTING_TOOL_ID)) {
    if (!weapon) {
      slots[WEAPON_TOOL_SLOT_ID] = { itemId: STARTING_HUNTING_TOOL_ID, quantity: 1 }
    } else {
      const existing = inventory.find((stack) => stack.itemId === STARTING_HUNTING_TOOL_ID)
      if (existing) existing.quantity += 1
      else inventory = [...inventory, { itemId: STARTING_HUNTING_TOOL_ID, quantity: 1 }]
    }
    return {
      ...save,
      inventory,
      equipment: { slots },
    }
  }

  // Already owned: equip it if the weapon/tool slot is empty.
  if (!weapon) {
    const invStack = inventory.find((stack) => stack.itemId === STARTING_HUNTING_TOOL_ID)
    if (invStack && invStack.quantity > 0) {
      inventory = inventory
        .map((stack) =>
          stack.itemId === STARTING_HUNTING_TOOL_ID
            ? { ...stack, quantity: stack.quantity - 1 }
            : stack,
        )
        .filter((stack) => stack.quantity > 0)
      slots[WEAPON_TOOL_SLOT_ID] = { itemId: STARTING_HUNTING_TOOL_ID, quantity: 1 }
      return {
        ...save,
        inventory,
        equipment: { slots },
      }
    }
  }

  return save
}

function remapRetiredNetId(itemId: string): string {
  return itemId === RETIRED_FISHING_NET_ITEM_ID ? STARTING_HUNTING_TOOL_ID : itemId
}

function stackMergeKey(stack: InventoryStack | EquippedStack): string {
  const enchantment = stack.enchantmentId ?? ''
  const favorite = stack.favorite === true ? '1' : '0'
  return `${stack.itemId}\0${enchantment}\0${favorite}`
}

function mergeRemappedStacks<T extends InventoryStack>(stacks: T[]): T[] {
  const merged: T[] = []
  const indexByKey = new Map<string, number>()
  for (const stack of stacks) {
    const next = { ...stack, itemId: remapRetiredNetId(stack.itemId) }
    const key = stackMergeKey(next)
    const existingIndex = indexByKey.get(key)
    if (existingIndex == null) {
      indexByKey.set(key, merged.length)
      merged.push(next)
      continue
    }
    const existing = merged[existingIndex]
    if (!existing) continue
    merged[existingIndex] = {
      ...existing,
      quantity: existing.quantity + next.quantity,
    }
  }
  return merged
}

const RETIRED_ITEM_IDS = new Set([
  'ITEM-0091', // Gem Setting
  'ITEM-0096', // Steel Spear Head
  'ITEM-0144', // Enchanted Sword
  'ITEM-0286', // Bone Shard
  'ITEM-0325', // Golden Spud Seed
  'ITEM-0346', // Hunting Trap
])

function isRetiredItemId(itemId: string | null | undefined): boolean {
  return typeof itemId === 'string' && RETIRED_ITEM_IDS.has(itemId)
}

function dropRetiredStacks<T extends InventoryStack>(stacks: T[]): T[] {
  return stacks.filter((stack) => !isRetiredItemId(stack.itemId))
}

function dropRetiredEquipped(stack: EquippedStack | null): EquippedStack | null {
  if (!stack || isRetiredItemId(stack.itemId)) return null
  return stack
}

/** Strip deleted leftover stacks so they cannot show as raw item IDs. */
export function removeRetiredItems(save: PlayerSave): PlayerSave {
  const slots: Record<string, EquippedStack | null> = {}
  for (const [slotId, stack] of Object.entries(save.equipment.slots)) {
    slots[slotId] = dropRetiredEquipped(stack)
  }
  const presets = (save.equipmentPresets ?? []).map((preset) => ({
    ...preset,
    slots: Object.fromEntries(
      Object.entries(preset.slots ?? {}).map(([slotId, stack]) => [
        slotId,
        dropRetiredEquipped(stack),
      ]),
    ),
  }))
  const locationTimers = (save.locationTimers ?? []).filter(
    (timer) => !isRetiredItemId(timer.inputItemId) && !isRetiredItemId(timer.outputItemId),
  )
  const lootTrackers = Object.fromEntries(
    Object.entries(save.lootTrackers ?? {}).map(([key, entry]) => [
      key,
      {
        ...entry,
        items: Object.fromEntries(
          Object.entries(entry.items ?? {}).filter(([itemId]) => !isRetiredItemId(itemId)),
        ),
      },
    ]),
  )
  return {
    ...save,
    inventory: dropRetiredStacks(save.inventory),
    bank: dropRetiredStacks(save.bank ?? []),
    equipment: { ...save.equipment, slots },
    equipmentPresets: presets,
    locationTimers,
    lootTrackers,
  }
}

/** Turn leftover Fishing Nets into the regular hunting Net. */
export function replaceFishingNetsWithNet(save: PlayerSave): PlayerSave {
  const slots: Record<string, EquippedStack | null> = {}
  for (const [slotId, stack] of Object.entries(save.equipment.slots)) {
    slots[slotId] = stack ? { ...stack, itemId: remapRetiredNetId(stack.itemId) } : null
  }
  return {
    ...save,
    inventory: mergeRemappedStacks(save.inventory),
    bank: mergeRemappedStacks(save.bank ?? []),
    equipment: { ...save.equipment, slots },
  }
}

File: packages/ik_rules/lib/src/save/starting_gear.dart
import '../js_compat.dart';
import 'generated/save_models.dart';
import 'json_save.dart';

bool _hasItem(List<Object?> inventory, SaveJson slots, String itemId) {
  final owned = inventory.any(
    (entry) =>
        entry is SaveJson && entry['itemId'] == itemId && isPositiveQuantity(entry['quantity']),
  );
  if (owned) return true;
  return slots.values.any(
    (stack) =>
        stack is SaveJson && stack['itemId'] == itemId && isPositiveQuantity(stack['quantity']),
  );
}

SaveJson? _firstStack(List<Object?> inventory, String itemId) {
  for (final entry in inventory) {
    if (entry is SaveJson && entry['itemId'] == itemId) return entry;
  }
  return null;
}

SaveJson _withGear(SaveJson save, List<Object?> inventory, SaveJson slots) {
  final next = copySave(save);
  next['inventory'] = inventory;
  next['equipment'] = <String, Object?>{'slots': slots};
  return next;
}

/// Ensure the demo hunting Net is owned and equipped when the tool slot is free.
///
/// Runs as a save migration, so it takes loose JSON: the save being upgraded is
/// only a version 4 save and has no guarantee of matching the current schema.
SaveJson ensureStartingHuntingToolJson(SaveJson save) {
  final slots = objectOrEmpty(objectAt(save, 'equipment') ?? <String, Object?>{}, 'slots');
  final inventory = arrayOrEmpty(save, 'inventory');
  final weapon = slots[weaponToolSlotId];

  if (!_hasItem(inventory, slots, startingHuntingToolId)) {
    if (weapon == null) {
      slots[weaponToolSlotId] = <String, Object?>{'itemId': startingHuntingToolId, 'quantity': 1};
    } else {
      final existing = _firstStack(inventory, startingHuntingToolId);
      if (existing != null) {
        existing['quantity'] = jsNumber(existing['quantity']) + 1;
      } else {
        inventory.add(<String, Object?>{'itemId': startingHuntingToolId, 'quantity': 1});
      }
    }
    return _withGear(save, inventory, slots);
  }

  // Already owned: equip it if the weapon/tool slot is empty.
  if (weapon == null) {
    final stack = _firstStack(inventory, startingHuntingToolId);
    if (stack != null && isPositiveQuantity(stack['quantity'])) {
      for (final entry in inventory) {
        if (entry is SaveJson && entry['itemId'] == startingHuntingToolId) {
          entry['quantity'] = jsNumber(entry['quantity']) - 1;
        }
      }
      inventory.removeWhere(
        (entry) => entry is! SaveJson || !isPositiveQuantity(entry['quantity']),
      );
      slots[weaponToolSlotId] = <String, Object?>{'itemId': startingHuntingToolId, 'quantity': 1};
      return _withGear(save, inventory, slots);
    }
  }

  return save;
}

/// Typed wrapper for callers that already hold a current-schema save.
PlayerSave ensureStartingHuntingTool(PlayerSave save) =>
    PlayerSave.fromJson(ensureStartingHuntingToolJson(save.toJson()));

String _remapRetiredNetId(Object? itemId) {
  return itemId == retiredFishingNetItemId ? startingHuntingToolId : jsString(itemId);
}

String _stackMergeKey(SaveJson stack) {
  final enchantment = stack['enchantmentId'] is String ? stack['enchantmentId'] as String : '';
  final favorite = stack['favorite'] == true ? '1' : '0';
  return '${stack['itemId']}\u0000$enchantment\u0000$favorite';
}

List<Object?> _mergeRemappedStacks(List<Object?> stacks) {
  final merged = <SaveJson>[];
  final indexByKey = <String, int>{};
  for (final entry in stacks) {
    if (entry is! SaveJson) continue;
    final next = SaveJson.of(entry);
    next['itemId'] = _remapRetiredNetId(next['itemId']);
    final key = _stackMergeKey(next);
    final existingIndex = indexByKey[key];
    if (existingIndex == null) {
      indexByKey[key] = merged.length;
      merged.add(next);
      continue;
    }
    final existing = merged[existingIndex];
    existing['quantity'] = jsNumber(existing['quantity']) + jsNumber(next['quantity']);
  }
  return merged;
}

/// Turns leftover Fishing Nets into the regular hunting Net.
SaveJson replaceFishingNetsWithNetJson(SaveJson save) {
  final next = copySave(save);
  next['inventory'] = _mergeRemappedStacks(arrayOrEmpty(save, 'inventory'));
  next['bank'] = _mergeRemappedStacks(arrayOrEmpty(save, 'bank'));
  final slots = objectOrEmpty(objectAt(save, 'equipment') ?? <String, Object?>{}, 'slots');
  final nextSlots = <String, Object?>{};
  for (final entry in slots.entries) {
    final stack = asObject(entry.value);
    if (stack == null) {
      nextSlots[entry.key] = entry.value;
      continue;
    }
    final remapped = SaveJson.of(stack);
    remapped['itemId'] = _remapRetiredNetId(remapped['itemId']);
    nextSlots[entry.key] = remapped;
  }
  next['equipment'] = <String, Object?>{'slots': nextSlots};
  return next;
}

PlayerSave replaceFishingNetsWithNet(PlayerSave save) =>
    PlayerSave.fromJson(replaceFishingNetsWithNetJson(save.toJson()));
